onnx/defs/op_annotation.h

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) Facebook Inc. and Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "onnx/common/assertions.h"
+
+namespace ONNX_NAMESPACE {
+
+struct EnumHash {
+  template <typename T>
+  std::size_t operator()(T t) const {
+    return static_cast<std::size_t>(t);
+  }
+};
+
+// Optional Annotation of an OpSchema which provides a level of generalization
+// to the optimization framework. These are simple properties of the Op in
+// question. Some annotations hierarchically contain other annotations. For
+// example an ElementwiseIndependent annotation necessarily contains
+// ElementwiseAny. This hierarchy is expressed in the OpAnnotation class.
+//
+// If you're adding an extra flag please remember to add it to the registry
+// below.
+enum class OpAnnotationFlag : uint16_t {
+  // Op is applied elementwise to the input. The output shape exactly matches
+  // the input shape.
+  ElementwiseAny,
+  // ElementwiseIndependent means the elementwise application of the Op does not
+  // rely on any other elements in that tensor. All unary mathematical offers
+  // are like this (e.g. Log, Exp, Sqrt).
+  ElementwiseIndependent,
+  // ElementwiseDependent is also an elementwise, but the elementwise dependent
+  // Op relies on information outside the current value. The softmax family of
+  // operators is a good example of an ElementwiseDependent Op. Although softmax
+  // is applied elementwise the partition portion of the calculation requires a
+  // summation across axis, therefore utilizing information outside of the
+  // current value.
+  ElementwiseDependent,
+  // Elementwise operation that is strictly monotonic increasing. In other words
+  // op(x) > op(y) iff x > y
+  ElementwiseStrictMonotonicIncreasing,
```

Comment:
Yeah, actually, I would like to ask whether we should make ElementwiseStrictMonotonicIncreasingOpAnn a subclass of ElementwiseWeakMonotonicIncreasingOpAnn? Since if you are strict, you must be weak too.